pkg/virt-handler/vm.go

```diff
@@ -2807,6 +2807,13 @@ func (d *VirtualMachineController) vmUpdateHelperDefault(origVMI *v1.VirtualMach
 
 		if err := d.hotplugVirtioInterfaces(vmi); err != nil {
 			log.Log.Object(vmi).Error(err.Error())
+			var criticalNetErr *neterrors.CriticalNetworkError
+			if !goerror.As(err, &criticalNetErr) {
+				log.Log.Object(vmi).Infof("pod interface not ready yet; re-queuing VMI reconcile operation to give it more time")
```

Comment:
Re-enqueuing is the `execute` default behavior in case an error is returned.
What are you trying to achieve here? Re-enqueuing only in case of a non critical error? In this case you should return an error only if the error is not critical (instead of re-enqueuing here).

In case of a critical error, how do you make sure the reconcile loop doesn't try to plug it again and again (each time the vmi gets somehow to the queue)? Only by the lock that is checked when stepping into the virt-launcher network namespace?

This change deserves its own commit with a clear commit message and a comment here explaining the behavior.

Reply:
These are the logs I see on the dynamic-networks-controller:
```
I0206 14:23:43.056223       1 pod.go:217] 1 attachments to add to pod kubevirt-test-default1/virt-launcher-testvmi-5mdhn-mvj97
I0206 14:23:43.081262       1 pod.go:236] 0 attachments to remove from pod kubevirt-test-default1/virt-launcher-testvmi-5mdhn-mvj97
I0206 14:23:43.082232       1 pod.go:144] extracted request [{"PodName":"virt-launcher-testvmi-5mdhn-mvj97","PodNamespace":"kubevirt-test-default1","AttachmentNames":[{"name":"linux-bridge-net-vlan100","namespace":"kubevirt-test-default1","interface":"net1"}],"Type":"add","PodNetNS":"/var/run/netns/ae031db0-efa9-4ce0-b669-b7d7c5bde09f"}] from the queue
I0206 14:23:43.082512       1 pod.go:152] handleDynamicInterfaceRequest: read from queue: {"PodName":"virt-launcher-testvmi-5mdhn-mvj97","PodNamespace":"kubevirt-test-default1","AttachmentNames":[{"name":"linux-bridge-net-vlan100","namespace":"kubevirt-test-default1","interface":"net1"}],"Type":"add","PodNetNS":"/var/run/netns/ae031db0-efa9-4ce0-b669-b7d7c5bde09f"}
I0206 14:23:43.082542       1 pod.go:252] network to add: &{linux-bridge-net-vlan100 kubevirt-test-default1 []   net1 [] <nil> <nil> []}
I0206 14:23:43.159805       1 pod.go:279] response: {1.0.0 [{Name:br10 Mac:7a:f9:70:70:2e:6c Sandbox:} {Name:vethd4c2e09c Mac:7a:f9:70:70:2e:6c Sandbox:} {Name:net1 Mac:de:c8:d3:49:c2:11 Sandbox:/var/run/netns/ae031db0-efa9-4ce0-b669-b7d7c5bde09f}] [] [] {[]  [] []}}
I0206 14:23:43.178976       1 event.go:285] Event(v1.ObjectReference{Kind:"Pod", Namespace:"kubevirt-test-default1", Name:"virt-launcher-testvmi-5mdhn-mvj97", UID:"9b708b53-a8e4-4a96-8f41-f6f6e8969130", APIVersion:"v1", ResourceVersion:"4674", FieldPath:""}): type: 'Normal' reason: 'AddedInterface' pod [kubevirt-test-default1/virt-launcher-testvmi-5mdhn-mvj97]: added interface net1 to network: linux-bridge-net-vlan100
```
The relevant thing here is the time: that particular layer spent 179 - 56 = 123 milliseconds to process adding the pod interface - this example uses bridge CNI under the hood - I guess other CNIs will take more time.

That's were I'm getting the 200 milliseconds from. If I re-queue using 5 / 25 / 625 /  ... I'm bound to fail the initial attempt, plus the 1st and 2nd retry (after 5ms and 25 ms respectfully).

200 milliseconds seemed a good enough compromise. I can also get rid of this, and let it fail 3 times for every request if you prefer ... it is indeed safer in the long term. 

> Maybe it worth moving the check that is done inside the virt-launcher network namespace to be done before stepping into the network namespace.

No, not worth it. Managing a state machine between controllers sounds like a bad idea - it is not the kubernetes way.

Reply:
I prefer the long term safer way.
I also don't understand why the first three attempts would fail? The virt-handler starts hot-plugging only after the cni finshed hot-plugging to the pod.